### PR TITLE
For #1499: add read up to status to chats

### PR DESCRIPTION
**Pull request by timothyclemansinsea**
To test: open several chrome user sessions, log into same project with different smc users, and type messages and trigger mark_read. You should see Avatars if < all project collaborators read message or only two collaborators or collaborators have different read up tos and "All collaborators read" otherwise. 

A sender's Avatar does not show if they've only read up to their last message.

image

image

### CODE REVIEW

File: src/smc-webapp/smc_chat.cjsx [hunk 1]
```diff
@@ -323,27 +323,105 @@ Message = rclass
                 {cols}
             </Row>
 
+Array::bisect = (val) ->
```
Comment (haraldschilly) [inline src/smc-webapp/smc_chat.cjsx]: could you please move this block of code from smc_chat to smc-util/misc ? it looks very generic and could be useful anywhere. besides that, I'm not sure if it is so good to change the prototype. `bisect` is only used in one place, and there is nothing gained by doing this. Just define a normal function and call it `array_bisect`.

Also, once it is in misc.coffee, add some tests to `smc-util/test/misc-test.coffee`

File: src/smc-webapp/smc_chat.cjsx [hunk 2]
```diff
@@ -362,6 +440,37 @@ ChatLog = rclass
                 else
                     @props.actions.set_editing(@props.messages.get(date), false)
 
+    find_read_tos: (read_dates_by_user, message_dates) ->
+        read_indexes_to_user_ids = {}
+        [user_ids, read_dates] = [Object.keys(read_dates_by_user), underscore.values(read_dates_by_user)]
+        find_between = (read_date) ->
```
Comment (haraldschilly) [inline src/smc-webapp/smc_chat.cjsx]: find_between appears twice. just make it its own function and use it here. makes it shorter and more readable :-)

File: src/smc-webapp/smc_chat.cjsx [hunk 3]
```diff
@@ -362,6 +440,37 @@ ChatLog = rclass
...
+            answer -= 1
+            if answer == -1
+                return null
+            return answer
+        indexes_for_read_up_to = (find_between(date) for date in read_dates)
+        for index_for_read_up_to, i in indexes_for_read_up_to
+            user_id = user_ids[i]
+            if index_for_read_up_to is null
+                continue
+            if String(index_for_read_up_to) in Object.keys(read_indexes_to_user_ids)
```
Comment (haraldschilly) [inline src/smc-webapp/smc_chat.cjsx]: maybe just `if index_for_read_up_to of read_indexes_to_user_ids` ?

Comment (haraldschilly) [inline src/smc-webapp/smc_chat.cjsx]: or even simpler, two lines:

```
read_indexes_to_user_ids.index_for_read_up_to ?= []
read_indexes_to_user_ids.index_for_read_up_to.push(user_id)
```

**Comment by timothyclemansinsea** [comment]
@haraldschilly I've cleaned this up per your review

File: src/smc-util/misc.coffee [hunk 4]
```diff
@@ -1898,3 +1898,17 @@ exports.bind_objects = (scope, arr_objects) ->
                 return bound_func
             else
                 return val
+
+# Provide the index that a value would be inserted
+# into an array where the val is between the number before
+# and after it
+exports.array_bisect = (arr, val) ->
```
Comment (johnjeng) [inline src/smc-util/misc.coffee]: Takes an Array and a Val. Returns the first index where the value there is greater than `Val`. Else returns the length of the array.

Or you can say something about assumptions of sortedness, but the code does no such thing.

File: src/smc-util/test/misc-test.coffee [hunk 5]
```diff
@@ -1389,3 +1389,8 @@ describe 'bind_objects', ->
 
         expect(b_obj1.func()).toEqual("cake")
         expect(b_obj1.val).toEqual("lies")
+
+describe 'array bisect function', ->
+    array_bisect = misc.array_bisect
+    it 'returns insertion index of 3 for a value of 11 for array of [0,3,5,15,30,90]', ->
+        array_bisect([0,3,5,15,30,90], 11).should.eql 3
```
Comment (johnjeng) [inline src/smc-util/test/misc-test.coffee]: Likewise, this description should be updated.

There should be tests for edge cases: Zero length array, smaller than all elements, larger than all elements.

File: src/smc-webapp/smc_chat.cjsx [hunk 6]
```diff
@@ -323,27 +323,86 @@ Message = rclass
                 {cols}
             </Row>
 
+sortCopy = (arr) ->
+    return arr.slice(0).sort()
+
+are_dicts_equal = (a, b) ->
```
Comment (johnjeng) [inline src/smc-webapp/smc_chat.cjsx]: I have to imagine this is in misc somewhere or use Underscore. Unless this does something special in which case that needs to be documented.

File: src/smc-webapp/smc_chat.cjsx [hunk 7]
```diff
@@ -323,27 +323,86 @@ Message = rclass
...
+    # assuming maps to lists
+    sorteda = {}
+    for key in sortCopy(Object.keys(a))
+        sorteda[key] = a[key]
+    sortedb = {}
+    for key in sortCopy(Object.keys(b))
+        sortedb[key] = b[key]
+    return JSON.stringify(sorteda) == JSON.stringify(sortedb)
+
+diffArray = (arr1, arr2) ->
```
Comment (johnjeng) [inline src/smc-webapp/smc_chat.cjsx]: Please comment what this does. Reading the code, it doesn't seem to do any of my first 3 guesses.

File: src/smc-webapp/smc_chat.cjsx [hunk 8]
```diff
@@ -323,27 +323,86 @@ Message = rclass
...
         set_scroll   : rtypes.func
 
+    getInitialState: ->
+        read_indexes_to_user_ids: @find_read_tos(@props.read_dates_by_user, @props.message_dates)
+        prev_read_dates_by_user: @props.read_dates_by_user
+
+    mixins: [SetIntervalMixin]
+
+    componentWillMount: ->
+        setInterval(@update_read_indexes_to_user_ids, 300)
```
Comment (johnjeng) [inline src/smc-webapp/smc_chat.cjsx]: I don't remember, does this automatically clear?

File: src/smc-webapp/smc_chat.cjsx [hunk 9]
```diff
@@ -323,27 +323,86 @@ Message = rclass
...
+
+    update_read_indexes_to_user_ids: ->
+        if are_dicts_equal(@props.read_dates_by_user, @state.prev_read_dates_by_user)
+            return
+        read_indexes_to_user_ids = @state.read_indexes_to_user_ids
+        # figure out which users to update
+        diff = diffArray(Object.keys(@props.read_dates_by_user), Object.keys(@state.prev_read_dates_by_user))
```
Comment (johnjeng) [inline src/smc-webapp/smc_chat.cjsx]: Why do we check if they're equal and then diff them? It seems we could diff them and based on that result know if they're equal or not.

File: src/smc-webapp/smc_chat.cjsx [hunk 10]
```diff
@@ -323,27 +323,86 @@ Message = rclass
...
+        for account_id in Object.keys(@props.read_dates_by_user)
+            if @props.read_dates_by_user[account_id] != @state.prev_read_dates_by_user[account_id]
+                if account_id not in diff
+                    diff.push(account_id)
```
Comment (johnjeng) [inline src/smc-webapp/smc_chat.cjsx]: I'm curious why this isn't covered by the diffing operation? It seems to rely on exactly the same data.

File: src/smc-webapp/smc_chat.cjsx [hunk 11]
```diff
@@ -323,27 +323,86 @@ Message = rclass
...
+
+    mixins: [SetIntervalMixin]
+
+    componentWillMount: ->
+        setInterval(@update_read_indexes_to_user_ids, 300)
+
+    update_read_indexes_to_user_ids: ->
+        if are_dicts_equal(@props.read_dates_by_user, @state.prev_read_dates_by_user)
+            return
+        read_indexes_to_user_ids = @state.read_indexes_to_user_ids
```
Comment (johnjeng) [inline src/smc-webapp/smc_chat.cjsx]: Does this make a copy?

File: src/smc-webapp/smc_chat.cjsx [hunk 12]
```diff
@@ -362,6 +421,36 @@ ChatLog = rclass
                 else
                     @props.actions.set_editing(@props.messages.get(date), false)
 
+    find_between: (read_date) ->
+        answer = misc.array_bisect(@props.message_dates, read_date)
+        answer -= 1
+        if answer == -1
+            return null
+        return answer
```
Comment (johnjeng) [inline src/smc-webapp/smc_chat.cjsx]: This seems to work around your design of `array_bisect`.

File: src/smc-webapp/smc_chat.cjsx [hunk 13]
```diff
@@ -406,6 +495,17 @@ ChatLog = rclass
                      set_scroll       = {@props.set_scroll}
                     />
 
+            if @state?.read_indexes_to_user_ids?[i]
+                if @state?.read_indexes_to_user_ids?[i].length > 0
```
Comment (johnjeng) [inline src/smc-webapp/smc_chat.cjsx]: When does `state` not exist? You have `getInitialState`.

Comment (johnjeng) [inline src/smc-webapp/smc_chat.cjsx]: Also, it shouldn't be two if statements.

File: src/smc-util/misc.coffee (hunk 4 above)
Comment (johnjeng) [inline src/smc-util/misc.coffee]: Also this doesn't seem to be doing any bisecting. It's more like the index finder for an insertion sort. Maybe name it `find_larger_entry`. This naming would be analogous to [Array.prototype.find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find). Which maybe this should use?

File: src/smc-webapp/smc_chat.cjsx (hunk 9 above)
Comment (timothyclemansinsea) [inline src/smc-webapp/smc_chat.cjsx]: diffArray does not work on associative arrays which is what `@props.read_dates_by_user, @state.prev_read_dates_by_user` are

File: src/smc-webapp/smc_chat.cjsx (hunk 11 above)
Comment (timothyclemansinsea) [inline src/smc-webapp/smc_chat.cjsx]: IDK

File: src/smc-webapp/smc_chat.cjsx (hunk 10 above)
Comment (timothyclemansinsea) [inline src/smc-webapp/smc_chat.cjsx]: because diffArray is only for arrays of values not associative arrays

**Comment by williamstein** [comment]
Can you move the read information to the right hand side, like @johnjeng showed us with Facebook's Messenger?

**Comment by timothyclemansinsea** [comment]
@williamstein now right aligned
image

**Comment by williamstein** [comment]
Just testing this for a few minutes, I feel like it is broken and just doesn't work right.  For example, with three users and trying things for a bit, then just letting it sit a few minutes looks like the following:

screen shot 2017-01-23 at 11 00 25 am

1. The heads should all be at the bottom exactly like with Messenger (that thing @johnjeng showed us).
2. There should be exactly two heads, or a message like "All messages seen by everybody" (whatever that think @johnjeng showed us does).    As it is, often there are no small heads at all -- just no information.
3. In the picture the position of the one small head that does appear is inconsistent between clients.
4. I didn't test, but what happens if there are 100 heads?  In SMC there are projects with 100+ collaborators on the same project and even opening the same chatroom (e.g., Kedlaya's class)...  Which heads get shown?  

Anyway, ether the file_use stuff is broken or unsuitable for this purpose, or whatever you have implemented has bugs.

**Comment by timothyclemansinsea** [comment]
@williamstein 
1. I just pulled up a large group chat in Facebook messenger on phone and see heads in various points, not all at bottom. I think there's an issue with auto scrolling not scrolling to the bottom. With Facebook messenger this model seems to make it easier to spot who hasn't read my last message.
2. Are you sure it's just not scrolled to bottom? Also as it's coded now if person last read right after they sent a message then I don't show their Avatar. Maybe I should only do this if its the last message.
For 3 and 4 I'll do more testing to figure out exactly what's going on.

**Comment by williamstein** [comment]
> I just pulled up a large group chat in Facebook messenger on phone and
see heads in various points, not all at bottom. I think there's an issue
with auto scrolling not scrolling to the bottom. With Facebook messenger
this model seems to make it easier to spot who hasn't read my last message

Thanks for checking -- you're right, I was confused.

I honestly have never used Facebook messenger or really anything with this
sort of "read notification" before, so I'm not sure what to expect.  I
think it is inconsistent in the screenshot I gave.

**Comment by timothyclemansinsea** [comment]
@williamstein OK studying the screenshot I do see the inconsistency. The immediate thing that comes to mind is the setinterval of 3000. I'll start testing in a few moments.

**Comment by williamstein** [comment]
> ... setinterval of 3000.

1. That isn't the reason, since I waited more than 3 seconds.
2. Why the heck are you using setInterval?  You should make rendering depend on the file use store, so whenever it changes (e.g., because a user marks something read), you can update the output accordingly.  There is almost never any reason to use setInterval in SMC.  If you're using it, probably you're doing things wrong.  Almost everything is event/data driven.

**Comment by timothyclemansinsea** [comment]
@williamstein I used setInterval to get around the can't change state during rendering issue. I need to save which index has been read by whom in order to efficiently modify the index to accounts mapping.

**Comment by williamstein** [comment]
> to get around the can't change state during rendering issue.

:man_shrugging:   If you're trying to get around that as an issue, you're surely not understanding redux/react...

**Comment by timothyclemansinsea** [comment]
Closing because I don't know react/redux enough to fix this
